finetune/util/group_metrics.py

```diff
@@ -0,0 +1,522 @@
+        }
+    return per_class_metrics
+
+def micro_avg(class_metrics):
+    """
+    Takes a dictionary of class metrics and calculates micro F1
+
+    :param class_metrics: A dictionary of class metrics as produced by
+    calc_class_metrics()
+    :return: Micro F1
+    """
+    TP = sum([c["true_positives"] for c in class_metrics.values()])
+    FP = sum([c["false_positives"] for c in class_metrics.values()])
+    FN = sum([c["false_negatives"] for c in class_metrics.values()])
+    precision = calc_precision(TP, FP)
+    recall = calc_recall(TP, FN)
+    return calc_f1(recall, precision)
+
+def macro_avg(class_metrics):
+    """
+    Takes a dictionary of class metrics and calculates macro F1
+
+    :param class_metrics: A dictionary of class metrics as produced by
+    calc_class_metrics()
+    :return: Macro F1
+    """
+    f1s = [c["f1_score"] for c in class_metrics.values()]
+    return np.average(f1s)
+
+def weighted_avg(class_metrics):
+    """
+    Takes a dictionary of class metrics and calculates weighted F1
+
+    :param class_metrics: A dictionary of class metrics as produced by
+    calc_class_metrics()
+    :return: Weighted F1
+    """
+    classes = list(class_metrics.keys())
+    f1s = [class_metrics[c]["f1_score"] for c in classes]
+    supports = [class_metrics[c]["support"] for c in classes]
+    return np.average(f1s, weight=supports)
+
+def get_average(average, class_metrics):
+    """
+    Calculates average F1 
+
+    :param average: A string, one of "micro", "macro", "weighted"
+    :param class_metrics: A dictionary of class metrics as produced by
+    calc_class_metrics()
+    :return: Average F1 as calculated by the appropriate function
+    """
+    fns = {
+        "micro": micro_avg,
+        "macro": macro_avg,
+        "weighted": weighted_avg
+    }
+    return fns[average.lower()](class_metrics)
+
+def group_metrics(preds, labels, span_type="exact", average=None):
+    """
+    Takes a set of group predictions and labels and returns either per-class
+    metrics or average metrics.
+
+    :param preds, labels: A list of list of groups. Output from joint grouping
+    models should be unzipped to seperate group and NER predictions before
+    being evaluated.
+    :param span_type: A string, one of "exact", "token", "overlap", "superset"
+    :param average: A string, one of "micro", "macro", "weighted"
+    :return: Average F1 if average was specified, else per-class metrics
+    """
+    per_class_metrics = calc_class_metrics(
+        preds, labels, metric_type="group", span_type=span_type
+    )
+    if average:
+        # Returns single F1 value
+        return get_average(average, per_class_metrics)
+    else:
+        return per_class_metrics
+
+def entity_in_group(entity, group):
+    """
+    Returns True if entity is contained in a group span, False otherwise
+    :param entity: A single entity span
+    :param group: A single group
+    """
+    for span in group["tokens"]:
```

Comment:
Use of "token" to refer to a group span is kind of confusing, makes me think these are BPEs or something.

Reply:
Yeah this is something that happened way early in the project and totally doesn't match what it is. Hopefully can find time to fix this today or tomorrow